File: apps/android/src/main/kotlin/space/bitos/app/data/settings/PrivacyPrefsStore.kt
package space.bitos.app.data.settings

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import space.bitos.core.settings.CommentPermission
import space.bitos.core.settings.MessagePermission
import space.bitos.core.settings.PrivacyPrefs
import space.bitos.core.settings.PrivacyPrefsContract

/**
 * APP-018a row 2 adapter: SharedPreferences-backed interaction-gate store
 * executing the shared [PrivacyPrefsContract]. Rules live in business-core;
 * this adapter persists and publishes.
 */
class PrivacyPrefsStore(
    private val prefs: SharedPreferences,
) {
    constructor(context: Context) : this(
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE),
    )

    private val _state = MutableStateFlow(load())
    val state: StateFlow<PrivacyPrefs> = _state.asStateFlow()

    fun reload() {
        _state.value = load()
    }

    fun update(transform: (PrivacyPrefs) -> PrivacyPrefs) {
        val next = transform(_state.value)
        prefs.edit().putString(KEY_WIRE, PrivacyPrefsContract.encode(next)).apply()
        _state.value = next
    }

    fun setMessagePermission(permission: MessagePermission) =
        update { it.copy(messagePermission = permission) }

    fun setCommentPermission(permission: CommentPermission) =
        update { it.copy(commentPermission = permission) }

    private fun load(): PrivacyPrefs =
        prefs.getString(KEY_WIRE, null)?.let(PrivacyPrefsContract::decode) ?: PrivacyPrefs()

    companion object {
        const val PREFS_NAME = "bitos_privacy"
        const val KEY_WIRE = "privacy_prefs_v1"
    }
}

File: apps/ios/BitOS/Platform/Settings/PrivacyPrefsStore.swift
import BusinessCore
import Foundation
import Observation

/**
 * APP-018a row 2 adapter (iOS): UserDefaults-backed interaction-gate store
 * executing the shared `PrivacyPrefsContract` through the bridge. Rules
 * live in business-core; this adapter persists and publishes.
 */
@MainActor
@Observable
final class PrivacyPrefsStore {
    static let storageKey = "privacy_prefs_v1"

    struct State: Equatable {
        var privateAccount = false
        var includeClientTag = true
        var activityVisible = true
        var readReceipts = true
        var sensitiveReason = true
        var storyShare = true
        var messagePermission = "everyone"   // followers | everyone | none
        var commentPermission = "everyone"   // everyone | followers | friends
    }

    private(set) var state = State()

    private let defaults: UserDefaults
    private let bridge = BusinessCoreBridge()

    init(defaults: UserDefaults? = UserDefaults(suiteName: "bitos.privacy")) {
        self.defaults = defaults ?? .standard
        reload()
    }

    func reload() {
        let wire = bridge.privacyPrefsDecode(json: defaults.string(forKey: Self.storageKey) ?? "")
        state = State(
            privateAccount: wire.privateAccount,
            includeClientTag: wire.includeClientTag,
            activityVisible: wire.activityVisible,
            readReceipts: wire.readReceipts,
            sensitiveReason: wire.sensitiveReason,
            storyShare: wire.storyShare,
            messagePermission: wire.messagePermission,
            commentPermission: wire.commentPermission
        )
    }

    func update(_ transform: (inout State) -> Void) {
        var next = state
        transform(&next)
        state = next
        persist()
    }

    private func persist() {
        let wire = BusinessCoreBridge.PrivacyPrefsWire(
            privateAccount: state.privateAccount,
            includeClientTag: state.includeClientTag,
            activityVisible: state.activityVisible,
            readReceipts: state.readReceipts,
            sensitiveReason: state.sensitiveReason,
            storyShare: state.storyShare,
            messagePermission: state.messagePermission,
            commentPermission: state.commentPermission
        )
        defaults.set(bridge.privacyPrefsEncode(wire: wire), forKey: Self.storageKey)
    }
}
